geopandas/sindex.py

```diff
@@ -22,21 +21,27 @@
 def has_sindex():
     """Dynamically checks for ability to generate spatial index.
     """
-    return get_sindex_class() is not None
+    try:
+        get_sindex_class()
+        return True
+    except RuntimeError:
+        return False
 
 
 def get_sindex_class():
     """Dynamically chooses a spatial indexing backend.
 
     Required to comply with _compat.USE_PYGEOS.
-    The selection order goes PyGEOS > RTree > None.
+    The selection order goes PyGEOS > RTree > Error.
     """
     if compat.USE_PYGEOS:
         return PyGEOSSTRTreeIndex
     if compat.HAS_RTREE:
         return RTreeIndex
-    warn("Spatial indexes require either `rtree` or `pygeos`.")
-    return None
+    raise RuntimeError(
```

Comment:
Shouldn't this be `ImportError` instead?

Reply:
I guess it could be. It could be a bit confusing since there's no import being done here...

Reply:
The substantive matter here is that there are unmet dependencies, which means that they could not be imported.  I don't know a better way to indicate that without subclassing `ImportError` as `MissingOptionalDependency` within geopandas, which seems a bit overboard...  (and in this case, the optional dependency is not actually optional)

I think it is OK to emit `ImportError` even though the user isn't directly importing something.  Pandas handles optional dependencies this way when you actually try to use a function that requires it:

```
>>> pd.read_feather('/tmp/test.feather')
...
ImportError: Missing optional dependency 'pyarrow'.  Use pip or conda to install pyarrow.
```

Reply:
Okay, if that's what Pandas is doing, it sounds fine to me!